test/integration/targets/apt/tasks/upgrade_autoremove.yml

```
@@ -0,0 +1,71 @@
+# https://github.com/ansible/ansible/issues/46314
+- block:
+    - name: Remove upgrades from the equation
+      apt:
+        upgrade: true
+        state: present
+        update_cache: true
+
+    - name: Install foobar, installs foo as a dependency
+      apt:
+        name: foobar=1.0.0
+        allow_unauthenticated: true
+
+    - name: Check foobar version
+      shell: dpkg -s foobar | grep Version | awk '{print $2}'
+      register: foobar_version
+
+    - name: Ensure the correct version of foobar has been installed
+      assert:
+        that:
+          - "'1.0.0' in foobar_version.stdout"
+
+    - name: test autoremove + upgrade (check mode)
```

This task upgrades foobar instead of removing the unused dependency foo, and changed is reported correctly with or without the patch. Writing tests before applying the patch can help ensure it will prevent a regression.

```suggestion
    - name: Remove foobar, leaving behind its dependency foo
      apt:
        name: foobar=1.0.0
        state: absent

    - name: test autoremove + upgrade (check mode)
```

---

Good catch. Thanks.